File: Python/algorithmic_toolbox/15_maximizing_salary.py
import sys


def find_largest_number_from_pieces(number_strings):
    class Number(object):
        def __init__(self, as_string):
            self.as_string = as_string

        def __lt__(n1, n2):
            number_1_as_string, number_2_as_string = n1.as_string, n2.as_string
            return number_1_as_string + number_2_as_string < number_2_as_string + number_1_as_string

    return "".join(sorted(number_strings, key=Number, reverse=True))


def main():
    data = sys.stdin.read().splitlines()
    number_strings = data[1].split()
    result = find_largest_number_from_pieces(number_strings)
    print(result)


if __name__ == "__main__":
    main()

File: R/algorithmic_toolbox/15_maximizing_salary.R
#!/usr/bin/env Rscript
Number <- function(n) {
    wrapper <- list(asString=n)
    wrapper
}

"[.number" <- function(ns, ix) {
    class(ns) <- "list"
    n <- ns[ix]
    class(n) <- "number"
    n
}

"==.number" <- function(w1, w2) {
    n1 <- w1[[1]]$asString
    n2 <- w2[[1]]$asString
    paste(n1, n2, sep = "") == paste(n2, n1, sep = "")
}

">.number" <- function(w1, w2) {
    n1 <- w1[[1]]$asString
    n2 <- w2[[1]]$asString
    paste(n1, n2, sep = "") > paste(n2, n1, sep = "")
}

AssembleLargestNumberFromPieces <- function(numberStrings) {
    numbers <- lapply(numberStrings, Number)
    class(numbers) <- "number"
    unname(unlist(sort(numbers, decreasing = TRUE)))
}

Main <- function() {
    conn <- file("stdin", "r")
    numberStrings <- as.list(as.integer(unlist(strsplit(readLines(conn, n = 2)[2], split = " "))))
    close(conn)
    result <- AssembleLargestNumberFromPieces(numberStrings)
    cat(paste0(result, collapse = ""), fill = TRUE)
}

if (!interactive()) {
    Main()
}
